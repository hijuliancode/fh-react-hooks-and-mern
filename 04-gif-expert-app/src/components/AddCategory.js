import React, { useState } from 'react'
import PropTypes from 'prop-types'

export const AddCategory = ({ setCategories }) => {
  const inputSearchElm = document.getElementById('searchCategory')
  const [ inputSearchValue, setInputSearchValue ] = useState('');

  const handleInputChange = (e) => {
    setInputSearchValue( e.target.value )
  }

  const handleSubmit = (e) => {
    e.preventDefault();

    if ( inputSearchValue.trim().length > 2 ) {
      let newCategory = {
        id: guidGenerator(),
        name: inputSearchValue
      }
      setCategories( cats => [newCategory, ...cats] )

      setInputSearchValue('')
      inputSearchElm.focus()
    }
  }

  function guidGenerator() { // Globally Unique IDentifier
    var S4 = function() {
       return (((1+Math.random())*0x10000)|0).toString(16).substring(1);
    };
    return (S4()+S4()+"-"+S4()+"-"+S4()+"-"+S4()+"-"+S4()+S4()+S4());
  }

  return (
    <form onSubmit={ handleSubmit }>
      <h2 className="mb-1 is-size-6">Add Category</h2>
      <div className="field has-addons mb-4 input-search">
        <div className="control">
          <input
            className="input"
            id="searchCategory"
            type="text"
            placeholder="Find a repository"
            value={ inputSearchValue }
            onChange={ handleInputChange } />
        </div>
        <div className="control">
          <button className="button is-info" type="submit">
            Search
          </button>
        </div>
      </div>
    </form>
  )
}

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired
}