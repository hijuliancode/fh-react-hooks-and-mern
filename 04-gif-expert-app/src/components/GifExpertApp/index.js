import React, { useState } from 'react';

const GifExpertAppComponent = () => {
  const [ categories, setCategories ] = useState([
    { id: 'OnePunch', name: 'One Punch'},
    { id: 'SamuraiX', name: 'Samurai X'},
    { id: 'DragonBall', name: 'Dragon Ball'},
  ])

  const handleAdd = () => {
    let inputSearch = document.getElementById('searchCategory')
    let inputSearchValue = inputSearch.value

    if ( inputSearchValue ) {
      let newCategorie = {
        id: guidGenerator(),
        name: inputSearchValue
      }
      setCategories( cats => [...cats, newCategorie] )

      // Reset Input
      inputSearch.value = ''
      inputSearch.focus()
    }
  }

  function guidGenerator() { // Globally Unique IDentifier
    var S4 = function() {
       return (((1+Math.random())*0x10000)|0).toString(16).substring(1);
    };
    return (S4()+S4()+"-"+S4()+"-"+S4()+"-"+S4()+"-"+S4()+S4()+S4());
}

  return (
    <div className="container">
      <h2 className="is-size-5-mobile is-size-4 has-text-centered mt-4">Gif Expert App</h2>
      <hr />

      <div className="field has-addons mb-4 input-search">
        <div className="control">
          <input className="input" id="searchCategory" type="text" placeholder="Find a repository" />
        </div>
        <div className="control">
          <button onClick={() => handleAdd()} className="button is-info" type="button">
            Search
          </button>
        </div>
      </div>

      {
        categories.map( (category) => {
          return <button key={category.id} className="button is-small is-primary is-outlined mr-1 mb-1">{category.name}</button>
        } )
      }

    </div>
  );
}
 
export default GifExpertAppComponent;