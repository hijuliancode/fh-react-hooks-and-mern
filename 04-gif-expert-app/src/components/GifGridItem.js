import React from 'react'
import PropTypes from "prop-types";

export const GifGridItem = ( { url, title, externalUrl } ) => {
  return (
    <div className="card animate__animated animate__fadeIn">
      <div className="card-image">
        <a href={ externalUrl } target="_blank" className="image is-4by3 animate__animated animate__fadeIn animate__fast" rel="noreferrer">
          <img src={ url } alt={ title } />
        </a>
      </div>
      <div className="card-content">
        <div className="media">
          <div className="media-content">
            <p className="title is-4">{ title }</p>
          </div>
        </div>
      </div>
    </div>
  )
}

GifGridItem.propTypes = {
  url: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  externalUrl: PropTypes.string.isRequired,
}