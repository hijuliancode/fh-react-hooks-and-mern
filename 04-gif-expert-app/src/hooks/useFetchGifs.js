import { useEffect, useState } from 'react'
import { getFetchGifs } from '../helpers/getFetchGifs'

export const useFetchGifs = ( category ) => {
  const [ state, setState ] = useState({
    data: [],
    loading: true
  })

  useEffect(() => {
    getFetchGifs(category)
      .then( result => {
        setState({
          data: result,
          loading: false
        })
      })
  }, [ category ])

  return state // { data: [], loading: true }
}