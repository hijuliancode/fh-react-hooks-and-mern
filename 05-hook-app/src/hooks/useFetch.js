import { useEffect, useState, useRef } from "react"

export const useFetch = ( url ) => {

  const isMounted = useRef(true)

  const [state, setState] = useState({
    data: null,
    error: null,
    loading: true,
  })

  useEffect(() => {
    return ( ) => { // cleanup
      // Cuando se remueve o limpia el componente cambia isMounted a false
      // así no se llamará de manera innecesaria el setState y no generará error ni fugas de memoria
      isMounted.current = false
    }
  }, [])

  useEffect(() => {

    setState({ // Reset
      data: null,
      error: null,
      loading: true,
    });

    fetch( url )
      .then( resp => resp.json() )
      .then( data => {
        if (isMounted.current) { // El componente esta montado, se puede llamar la data
          setState({
            data: data[0],
            error: null,
            loading: false,
          })
        } else {
          console.log('Componente no montado no es necesario llamar data')
        }
      })
    
  }, [url])

  return state
}
