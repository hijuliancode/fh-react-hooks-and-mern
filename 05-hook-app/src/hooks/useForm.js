import { useState } from "react"

export const useForm = (initialState = {}) => {
  const [ values, setValues ] = useState(initialState)

  const handleInputChante = ({target}) => {
    setValues({
      ...values,
      [target.name]: [target.value]
    })
  }

  return [ values, handleInputChante ]
}
