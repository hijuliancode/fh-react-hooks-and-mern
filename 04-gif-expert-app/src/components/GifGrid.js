import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import { useFetchGifs } from '../hooks/useFetchGifs'
import { GifGridItem } from './GifGridItem'

export const GifGrid = ({ category }) => {

  const { data:images, loading } = useFetchGifs(category)

  return (
    <>
      { // If is loading
        loading && (
          <div class="notification animate__delay-1s is-success animate__animated animate__flash animate__infinite">
            <p>Cargando...</p>
          </div>
        )
      }
      <h2 className="title is-5 is-capitalized">{ category }</h2>
      <div className="columns is-multiline is-tablet mb-6">
        {
          images.map( image => (
            <div key={ image.id } className="column is-one-quarter">
              <GifGridItem { ...image } />
            </div>
          ))
        }
      </div>
    </>
  )
}

GifGrid.propTypes = {
  category: PropTypes.string.isRequired
}