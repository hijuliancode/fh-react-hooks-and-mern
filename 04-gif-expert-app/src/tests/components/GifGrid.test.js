import React from 'react'
import '@testing-library/jest-dom'
import { shallow } from 'enzyme'
import { GifGrid } from '../../components/GifGrid'
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs')


describe('Test in GifGrid', () => {
  let wrapper;

  test('Should render <GifGid />', () => {
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true
    })
    wrapper = shallow( <GifGrid category='Thor' /> )
    expect( wrapper ).toMatchSnapshot()
  })

  test('Debe mostrar items cuando se cargan imagenes con useFetchGifs', () => {
    const gifs = [
        {
        id: 'LSJ12SAKDJS2',
        url: 'https://cualquiercosa.com/demo-imagen.png',
        externalUrl: 'https://cualquiercosa.com/demo-imagen.png',
        title: 'Cualquier cosa'
      },
      {
        id: 'FDAHWR$%DFAF',
        url: 'https://cualquiercosa.com/demo-imagen-2.png',
        externalUrl: 'https://cualquiercosa.com/demo-imagen-2.png',
        title: 'Cualquier otra cosa'
      }
    ]

    useFetchGifs.mockReturnValue({
      data: gifs,
      loading: false
    })
    wrapper = shallow( <GifGrid category='Thor' /> )

    expect( wrapper.find('p').exists() ).toBe(false)
    expect( wrapper.find('GifGridItem').length ).toBe( gifs.length )
  })
})
