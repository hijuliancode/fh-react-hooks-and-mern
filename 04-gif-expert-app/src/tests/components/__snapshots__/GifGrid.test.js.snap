// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in GifGrid Should render <GifGid /> 1`] = `
<Fragment>
  <div
    className="notification animate__delay-1s is-success animate__animated animate__flash animate__infinite"
  >
    <p>
      Cargando...
    </p>
  </div>
  <h2
    className="title is-5 is-capitalized"
  >
    Thor
  </h2>
  <div
    className="columns is-multiline is-tablet mb-6"
  />
</Fragment>
`;
